srml/bridge/src/lib.rs

```diff
@@ -0,0 +1,257 @@
+			block_hash: &T::Hash,
+			state_root: &T::Hash,
+			validator_set: Vec<T::ValidatorId>,
+		) -> Self
+	{
+		// I don't like how this is done, should come back to...
+		BridgeInfo {
+			last_finalized_block_number: *block_number,
+			last_finalized_block_hash: *block_hash,
+			last_finalized_state_root: *state_root,
+			current_validator_set: validator_set,
+		}
+	}
+}
+
+type BridgeId = u64;
+
+pub trait Trait: system::Trait {
+	/// A stable ID for a validator.
+	type ValidatorId: Member + Parameter;
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as Bridge {
+		/// The number of current bridges managed by the module.
+		pub NumBridges get(num_bridges) config(): BridgeId;
+
+		/// Maps a bridge id to a bridge struct. Allows a single
+		/// `bridge` module to manage multiple bridges.
+		pub TrackedBridges get(tracked_bridges): map BridgeId => Option<BridgeInfo<T>>;
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		// TODO: Figure out the proper type for these proofs
+		fn initialize_bridge(
+			origin,
+			block_header: T::Header,
+			validator_set: Vec<T::ValidatorId>,
+			validator_set_proof: Vec<u8>,
+			storage_proof: Vec<u8>,
+		) {
+			// NOTE: Will want to make this a governance issued call
+			let _sender = ensure_signed(origin)?;
+
+			Self::check_storage_proof(storage_proof)?;
+			Self::check_validator_set_proof(validator_set_proof)?;
+
+			let block_number = block_header.number();
+			let block_hash = block_header.hash();
+			let state_root = block_header.state_root();
+			let bridge_info = BridgeInfo::new(block_number, &block_hash, state_root, validator_set);
+
+			let new_bridge_id = NumBridges::get() + 1;
+
+			// Hmm, can a call to `insert` fail?
+			// If this is an Option, why do we not need to wrap it in Some(T)?
+			<TrackedBridges<T>>::insert(new_bridge_id, bridge_info);
+
+			// Only increase the number of bridges if the insert call succeeds
+			NumBridges::put(new_bridge_id);
+		}
+
+		fn submit_finalized_headers(origin) {
+			let _sender = ensure_signed(origin)?;
+		}
+	}
+}
+
+decl_error! {
+	// Error for the Bridge module
+	pub enum Error {
+		InvalidStorageProof,
+		InvalidValidatorSetProof,
+	}
+}
+
+impl<T: Trait> Module<T> {
+	fn check_storage_proof(_proof: Vec<u8>) -> std::result::Result<(), Error> {
+		// ... Do some math ...
+		Ok(()) // Otherwise, Error::InvalidStorageProof
+	}
+
+	fn check_validator_set_proof(_proof: Vec<u8>) -> std::result::Result<(), Error> {
+		// ... Do some math ...
+		Ok(()) // Otherwise, Error::InvalidValidatorSetProof
+	}
+}
+
+#[cfg(test)]
+mod tests {
+	use super::*;
+
+	use primitives::{H256, Blake2Hasher};
+	use sr_primitives::{
+		Perbill, traits::{Header as HeaderT, IdentityLookup}, testing::Header, generic::Digest,
+	};
+	use support::{assert_ok, impl_outer_origin, parameter_types};
+	use runtime_io::with_externalities;
+
+	// NOTE: What's this for?
+	impl_outer_origin! {
+		pub enum Origin for Test {}
+	}
```

Comment:
You can define custom origins, so there is an outer origin which wraps the different origins you support.